### PR TITLE
HARP-5398: Stable labels across zoom levels

### PR DESCRIPTION
**Pull request by pit-rpg**
Signed-off-by: Stepan Krovspei <11714928+pit-rpg@users.noreply.github.com>

### CODE REVIEW

File: @here/harp-mapview/lib/text/TextElement.ts [hunk 1]
```diff
@@ -609,4 +612,21 @@ export class TextElement {
             }
         }
     }
+
+    /**
+     * Something like a unique hash for each text element, the text element position, and uid are
+     * ignored
+     */
+    getIdentityKey(): string {
```
Comment (ninok) [inline @here/harp-mapview/lib/text/TextElement.ts]: As far as I can see the unique id is mainly based on the text. What if the label has no text (I guess this can happen for POIs) or even worse if the text is the same (this will happen for sure for POIs like gas stations).

Comment (pit-rpg) [reply @here/harp-mapview/lib/text/TextElement.ts]: I can't identify labels between zoom levels because their position changes at each level.
This function generates a key by which I can place similar labels render states in one array, and then if needed we can find and copy a state for a new label at another zoom level.
I'll try to rename this method

File: @here/harp-mapview/lib/text/TextElement.ts [hunk 2]
```diff
@@ -609,4 +612,20 @@ export class TextElement {
             }
         }
     }
+
+    /**
+     * Something like a hash for similar text elements, the text element position, and uid ignored
```
Comment (haraldF) [inline @here/harp-mapview/lib/text/TextElement.ts]: "Something like" is not good documentation style. Can you reword please?

File: @here/harp-mapview/lib/text/TextElement.ts [hunk 3]
```diff
@@ -291,8 +291,25 @@ export class RenderState {
...
+            this.startTime,
+            this.opacity,
+            this.lastFrameNumber,
+            this.fadingTime
+        );
+    }
 }
 
+// Unique ID counter
+let TEXT_ELEMENT_UID = 0;
```
Comment (okanishchev) [inline @here/harp-mapview/lib/text/TextElement.ts]: It would be better to make it as private static property of TextElement class:

```js
class TextElement {
    private static counter: number = 0;
    readonly uid: number = ++TextElementCounter;
}
```

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts [hunk 4]
```diff
@@ -280,13 +306,49 @@ export class TextElementsRenderer {
...
+        // if level do not loaded or old data cleaned up do nothing
+        if (!zoomLevelCompletelyLoaded || this.m_renderStatesCleanedUp) {
+            return;
+        }
+
+        // if level completely loaded clean up unnecessary states
+        for (let [key, textElementStates] of Object.entries(this.m_renderStateStore)) {
+            // states that older then 120 frames are removed
+            textElementStates = textElementStates.filter(
+                state => state.renderStateText.lastFrameNumber + 120 > frameNumber
```
Comment (okanishchev) [inline @here/harp-mapview/lib/text/TextElementsRenderer.ts]: Why 120 is used? Saving this as constant with a descriptive name somewhere at the top and referencing here will be more readable.

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts [hunk 5]
```diff
@@ -858,17 +906,15 @@ export class TextElementsRenderer {
...
         // than once per element. Also, make sorting stable by taking the index into the array into
         // account, this is required to get repeatable results for testing.
-        for (let i = 0; i < textElements.length; i++) {
-            const textElement = textElements[i];
-
+        for (const textElement of textElements) {
             textElement.sortPriority =
                 textElement.priority +
-                i * indexPriorityFactor +
+                textElement.priority * uidPriorityFactor +
```
Comment (okanishchev) [inline @here/harp-mapview/lib/text/TextElementsRenderer.ts]: I believe this is a mistake
```suggestion
                textElement.uid * uidPriorityFactor +
```

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts [hunk 6]
```diff
@@ -2232,32 +2242,27 @@ export class TextElementsRenderer {
...
-            if (renderState.isFading()) {
-                this.updateFading(renderState, time);
-                fadeAnimationStarted = true;
-            }
-
-            renderState.lastFrameNumber = frameNumber;
+    // if not fading in fade in, if fading out reverse
+    private ensureFadeIn(renderState: RenderState, frameNumber: number, time: number) {
+        renderState.lastFrameNumber = frameNumber;
+        if (renderState.state === FadingState.FadingIn) {
```
Comment (okanishchev) [inline @here/harp-mapview/lib/text/TextElementsRenderer.ts]: It's a perfect case to use ```switch```

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts [hunk 7]
```diff
@@ -2232,32 +2242,27 @@ export class TextElementsRenderer {
...
-            }
-
-            if (renderState.isFading()) {
-                this.updateFading(renderState, time);
-                fadeAnimationStarted = true;
-            }
-
-            renderState.lastFrameNumber = frameNumber;
+    // if not fading in fade in, if fading out reverse
+    private ensureFadeIn(renderState: RenderState, frameNumber: number, time: number) {
```
Comment (okanishchev) [inline @here/harp-mapview/lib/text/TextElementsRenderer.ts]: The function named ensureFadeIn returns false if render state is FadedIn and true otherwise. And that is completely non-intuitive.
I would prefer to preserve that variable ```fadeAnimationStarted``` to keep awareness on what's going on.

File: @here/harp-mapview/lib/text/TextElement.ts [hunk 8]
```diff
@@ -609,4 +613,21 @@ export class TextElement {
             }
         }
     }
+
+    /**
+     * Generates a string from text object properties that can be used to find similar elements.
+     * The text element position and uid ignored
```
Comment (musculman) [inline @here/harp-mapview/lib/text/TextElement.ts]: Perhaps it is missing "are ignored"? instead of just "ignored"

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts (hunk 4 above)
Comment (musculman) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: Yes please, say no to magic numbers :)

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts [hunk 9]
```diff
@@ -858,17 +906,15 @@ export class TextElementsRenderer {
 
     private sortTextElements(textElements: TextElement[], maxViewDistance: number) {
         const distancePriorityFactor = 0.1;
-        const indexPriorityFactor = 0.01 * (1 / textElements.length);
+        const uidPriorityFactor = 0.0000000001;
```
Comment (musculman) [inline @here/harp-mapview/lib/text/TextElementsRenderer.ts]: Why this change from previous formula to hard-coded value?

Comment (pit-rpg) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: The previous one calculates the priority used the array length, but, the length is changing and this affects the priority across frames. To make the priority value more stable I've used the text element UID. So if the visible tileset is changed during camera movement this does not affect the displayed label's priority.
This value just a very small but static number for reducing the UID weight on text element priority.

**Comment by pit-rpg** [comment]
Decoders tests please

**Comment by DimitryRd** [comment]
Decoders tests please
